### DISCUSSION TITLE
Having trouble marking notifications as read for polymorphic comments

### DISCUSSION THREAD

**Discussion by DebakeDSP94** [excid3/noticed#278] [Q&A]
Hi.  Please forgive my newness to rails.  I currently learning from The Odin Project.  I'm using the noticed gem in my facebook clone app. I plan to eventually add nested comments, and notifications for likes and friend requests, but I want to get them working properly on comments first and then move forward.  The notifications are working, and are correctly linking to the post as intended.  The issue I'm having is in marking them as read.  I followed this tutorial, but had to change some things to have it work with polymorphic comments, and I think that's where my issue is. https://youtu.be/c570-wnW_-A

`  def mark_notifications_as_read
    if current_member
      notifications_to_mark_as_read = @post.notifications_as_post.where(recipient: current_member)
      notifications_to_mark_as_read.update_all(read_at: Time.zone.now)
    end
  end`

This snippet of code does not seem to be working.  I had thought that it should be modified to be @commentable, but when I tried that I got a no method error.  As it stands, when I click on a notification it is running an UPDATE "notifications" SET "read_at" in the rails server output, so it looks like it should be working, but when I look in irb, read_at is still nil for the notification.  I have posted the full question on stackoverflow here: https://stackoverflow.com/questions/75914440/rails-notifications-with-noticed-gem-not-getting-marked-as-read-on-polymorphic-c .I have not received any responses, so I'm kind of out of ideas at the moment.  No google searches I've done have provided results for using the noticed gem with polymorphic entities.  So, if you have any ideas of what I should try, then I would be really thankful.  Thank you so much for your great gem.  Please let me know if there is any more info I can provide to help find an answer.

**Comment by excid3** [comment]
Hey @DebakeDSP94 !

It looks like your query for notification is what's wrong `@post.notifications_as_post` is only returning notifications that reference the Post, but a nested comment would probably only reference the parent comment. You will want to adjust that query to return the notifications that reference the comment to handle that case.

**Reply by DebakeDSP94** [reply]
Thanks Chris.  By the way, this is the guy from reddit that commented on your april fools post.  So, I'm trying to figure out how exactly to do what you suggest.  In the rails console, if I look at a notification, this is what I get
#<Notification:0x00007f237ac8e6d8
  id: 2,
  recipient_type: "Member",
  recipient_id: 1,
  type: "CommentNotification",
  params:
   {:post=>nil,
    :comment=>
     #<Comment:0x00007f237ac6e888
      id: 2,
      member_id: 2,
      commentable_type: "Post",
      commentable_id: 1,
      body: nil,
      created_at: Mon, 03 Apr 2023 20:07:47.836134000 UTC +00:00,
      updated_at: Mon, 03 Apr 2023 20:07:47.840325000 UTC +00:00>},
  read_at: nil,
  created_at: Mon, 03 Apr 2023 20:07:47.869464000 UTC +00:00,
  updated_at: Mon, 03 Apr 2023 20:07:47.869464000 UTC +00:00>]

I've tried a few variations of @commentable.comment.notifications.where(recipent: current_member), but every time I'm getting a n error of:
undefined method `comment' for #<ActiveRecord::Associations::CollectionProxy [#<Comment id: 1, member_id: 2, commentable_type: "Post", commentable_id: 1, body: nil, created_at: "2023-04-02 19:48:51.167845000 +0000", updated_at: "2023-04-02 19:48:51.171893000 +0000">, #<Comment id: 2, member_id: 2, commentable_type: "Post", commentable_id: 1, body: nil, created_at: "2023-04-03 20:07:47.836134000 +0000", updated_at: "2023-04-03 20:07:47.840325000 +0000">]>  
So, I'm still struggling with it a little on how exactly I should structure it.

**Reply by DebakeDSP94** [reply]
I guess what I'm really not understanding is what the "notifications_as_post" method is doing.  I can use @ post.notifications.where(recipient: current_member) and that just instantly marks everything as read and turns the notification count to 0.  If I do the @post.notifications_as_post.where(recipient: current_member) as per the tutorial I followed, then there's no error, but nothing gets marked as read.  So, I feel like I'm fundamentally missing something important here.  I also wonder if I should be moving this function to the comment controller or maybe adding a before action of set_commentable so that maybe the function would work with @commentable vs @post.  I would appreciate any thoughts you have.  I'm sure that something important is just not clicking yet.  Thanks for your time.
